clang-tools-extra/test/clang-tidy/checkers/bugprone/unsafe-crtp.cpp

```diff
@@ -0,0 +1,232 @@
+// RUN: %check_clang_tidy %s bugprone-unsafe-crtp %t
+
+namespace class_implicit_ctor {
+template <typename T>
+class CRTP {};
+// CHECK-MESSAGES: :[[@LINE-1]]:7: warning: the implicit default constructor of the CRTP is publicly accessible [bugprone-unsafe-crtp]
+// CHECK-MESSAGES: :[[@LINE-2]]:7: note: consider making it private
+// CHECK-FIXES: CRTP() = default;
+
+class A : CRTP<A> {};
+} // namespace class_implicit_ctor
+
+namespace class_uncostructible {
+template <typename T>
+class CRTP {
+// CHECK-MESSAGES: :[[@LINE-1]]:7: warning: the CRTP cannot be constructed from the derived class [bugprone-unsafe-crtp]
+// CHECK-MESSAGES: :[[@LINE-2]]:7: note: consider declaring the derived class as friend
+// CHECK-FIXES: friend T;
```

Comment:
boost::noncopyable may not be best example, but thing is that private constructor is not an issue, simply check shouldn't say, "hey you forget friend" because such class may have private constructor on purpose, for example to use an factory function or just to provide static methods.

Reply:
I don't see how a factory function in a CRTP context would make sense. If the only constructor is for a static factory function, which is unaccessible by the derived class, then the derived class cannot be constructed either, right? Also since the goal of the CRTP is to provide some functionality to it's template parameter, having a factory method would just result in instantiating it on it's own, which also doesn't make sense, right?

Otherwise since the goal is to prevent any kind of accidental instantiation, the check assumes that every other constructor is private, and then we need the friend, since the derived class has no way to instantiate the CRTP. If anything else other than the derived class and the CRTP itself has access to the constructors, then accidental instantiation can happen, which we want to prevent.

If the class only provides static methods, then why would any other class inherit from it, or use it as a CRTP?

Reply:
> If the only constructor is for a static factory function, which is unaccessible by the derived class, then the derived class cannot be constructed either, right?

@isuckatcs I think you are right, at least I tried to give this some thought (actually by trying to construct a counter-example).

```cpp
template <class T>
class Base
{
    Base(int I) {}

protected:
    static Base factory(int I) { return Base{I}; }
};

class Derived : public Base<Derived>
{
    // Derived() : Base(0) {} // Does not work.
    Derived() : Base(Base::factory(1)) {}
};

void test()
{
    Derived D;
}
```

```
<source>:18:13: error: calling a private constructor of class 'Derived'
   18 |     Derived D;
      |             ^
<source>:13:5: note: implicitly declared private here
   13 |     Derived() : Base(Base::factory(1)) {}
      |     ^
```

Reply:
@whisperity I think you meant `struct Derived`, and then you example proves me wrong.

The check should also warn about the accessibility of any function, which can construct the CRTP from anywhere other than the derived class and itself.

Reply:
Since checking if a public or protected function can construct the class would require path sensitive analysis I'd add this example as a limitation to the docs.

E.g.:
```c++
template <class T>
class CRTP
{
    CRTP(int I) {}

public:
    static void hardToDetect(CRTP* Escape, int X) { 
        if(X > 2) {
            Escape = new CRTP{1};
        }
    }
};
```

Another solution would be to walk the callgraph, but that would just pollute the user with non-meaningful warnings.

E.g.:
```c++
template <class T>
class CRTP
{
    CRTP(int I) {}

public:
    void hardToDetect(CRTP* Escape, int X) { 
        if(CRTP && X > 2) {
            Escape = new CRTP{1};
        }
    }
    
    void warning1() { hardToDetect(nullptr, 3); }
    void warning2() { warning1(); }
    void warning3() { warning2(); }
};
```
In the above `CRTP` we could report a warning for every function, right?